First Set of Models

Billiet-mc opened the pull request:
new file:   backend/app/Models/Cache.php
	new file:   backend/app/Models/CacheLock.php
	new file:   backend/app/Models/DegreeProgram.php
	new file:   backend/app/Models/Department.php
	new file:   backend/app/Models/Faculty.php
	new file:   backend/app/Models/Job.php
	new file:   backend/app/Models/JobBatch.php
	new file:   backend/app/Models/Organization.php
	new file:   backend/app/Models/Session.php
	new file:   backend/app/Models/Student.php
	modified:   backend/app/Models/User.php

Billiet-mc committed:
new file:   backend/app/Models/Admin.php
	new file:   backend/app/Models/Cache.php
	new file:   backend/app/Models/CacheLock.php
	new file:   backend/app/Models/DegreeProgram.php
	new file:   backend/app/Models/Department.php
	new file:   backend/app/Models/Faculty.php
	new file:   backend/app/Models/Job.php
	new file:   backend/app/Models/JobBatch.php
	new file:   backend/app/Models/Organization.php
	new file:   backend/app/Models/Session.php
	new file:   backend/app/Models/Student.php
	modified:   backend/app/Models/User.php

Mephisto-22 commented on backend/app/Models/Cache.php:
This file can be removed, it was auto generated

Mephisto-22 commented on backend/app/Models/CacheLock.php:
This file can also be removed

Mephisto-22 commented on backend/app/Models/User.php:
```diff
@@ -1,50 +1,72 @@
+		'email_verified_at',
+		'remember_token'
+	];
+
```
Double check all these relationships to corespond to documentation

Mephisto-22 commented on backend/app/Models/Session.php:
This can also be deleted

Mephisto-22 commented on backend/app/Models/JobBatch.php:
This also can be deleted.

Mephisto-22 commented on backend/app/Models/Job.php:
Delete this file

Mephisto-22 commented on backend/app/Models/Faculty.php:
```diff
@@ -0,0 +1,66 @@
+		return $this->hasMany(Student::class);
+	}
+
+	public function departments(): HasMany
```
This needs to be a hasOne. A faculty can only be chair of one department

Mephisto-22 commented on backend/app/Models/Faculty.php:
```diff
@@ -0,0 +1,66 @@
+
+	public function degree_programs(): HasMany
+	{
+		return $this->hasMany(DegreeProgram::class, 'program_chair');
```
This also needs to be hasOne. A faculty cannot be the chair of more than one program

Mephisto-22 commented on backend/app/Models/Department.php:
```diff
@@ -0,0 +1,61 @@
+
+	public function faculty(): BelongsTo
+	{
+		return $this->belongsTo(Faculty::class, 'department_chair');
```
Change to hasOne for department chair.

Mephisto-22 commented on backend/app/Models/DegreeProgram.php:
```diff
@@ -0,0 +1,52 @@
+
+	public function faculty(): BelongsTo
+	{
+		return $this->belongsTo(Faculty::class, 'program_chair');
```
change to hasOne

Billiet-mc committed:
deleted:    backend/app/Models/Cache.php
	deleted:    backend/app/Models/CacheLock.php
	modified:   backend/app/Models/DegreeProgram.php
	modified:   backend/app/Models/Department.php
	modified:   backend/app/Models/Faculty.php
	deleted:    backend/app/Models/Job.php
	deleted:    backend/app/Models/JobBatch.php
	deleted:    backend/app/Models/Session.php
	modified:   backend/app/Models/User.php

BradyBurns028 commented on backend/app/Models/DegreeProgram.php:
```diff
@@ -0,0 +1,52 @@
+		return $this->belongsTo(Department::class);
+	}
+
+	public function faculty(): HasOne
```
This relationship should be called programChair and it belongsTo to a faculty. Note here how department is belongsTo and the degree_program has a department_id just as it does a program_chair.

BradyBurns028 commented on backend/app/Models/Department.php:
```diff
@@ -0,0 +1,61 @@
+		return $this->belongsTo(Organization::class);
+	}
+
+	public function faculty(): HasOne
```
Same as previous comment. departmentChair and belongsTo. The goal is that I can get $department->departmentChair to get the user that is that relationship. Getting $department->faculty is less intuitive.

BradyBurns028 commented on backend/app/Models/Department.php:
```diff
@@ -0,0 +1,61 @@
+		return $this->hasMany(DegreeProgram::class);
+	}
+
+	public function faculties(): HasMany
```
And keeping with intuition. I would probably call this simply faculty as this is the proper plural.

BradyBurns028 commented on backend/app/Models/Faculty.php:
```diff
@@ -0,0 +1,66 @@
+		return $this->belongsTo(Department::class);
+	}
+
+	public function degree_programs(): HasOne
```
Should be camelCase degreePrograms and it is a HasMany relationship as there is no prevention of one faculty member being a chair of more than one department.

BradyBurns028 commented on backend/app/Models/Faculty.php:
```diff
@@ -0,0 +1,66 @@
+		return $this->hasOne(DegreeProgram::class, 'program_chair');
+	}
+
+	public function students(): HasMany
```
Again following intuition, are these students or advisees?

BradyBurns028 commented on backend/app/Models/Faculty.php:
```diff
@@ -0,0 +1,66 @@
+		return $this->hasMany(Student::class);
+	}
+
+	public function departments(): HasOne
```
Can also be more than one HasMany

BradyBurns028 commented on backend/app/Models/User.php:
```diff
@@ -1,50 +1,72 @@
+ *
+ * @package App\Models
+ */
+class User extends Model
```
We broke Authenticable for this model. Should be class User Extends Authenticable

BradyBurns028 commented on backend/app/Models/User.php:
```diff
@@ -1,50 +1,72 @@
+use Carbon\Carbon;
+use Illuminate\Database\Eloquent\Collection;
+use Illuminate\Database\Eloquent\Model;
+use Illuminate\Database\Eloquent\SoftDeletes;
```
add use Illumintate\Foundation\Auth\User as Authenticable;

BradyBurns028 commented on backend/app/Models/User.php:
```diff
@@ -1,50 +1,72 @@
-            'password' => 'hashed',
-        ];
-    }
+	use SoftDeletes;
```
need additional traits of HasApiTokens and Notifiable

BradyBurns028 commented on backend/app/Models/Student.php:
```diff
@@ -0,0 +1,50 @@
+		'faculty_id'
+	];
+
+	public function degree_program(): BelongsTo
```
camel case

Billiet-mc committed:
modified:   backend/app/Models/DegreeProgram.php
	modified:   backend/app/Models/Department.php
	modified:   backend/app/Models/Faculty.php
	modified:   backend/app/Models/Student.php
	modified:   backend/app/Models/User.php
